From fed7f328c5a01a8b9afb7beea11a1676aee8badf Mon Sep 17 00:00:00 2001
From: Julian Bouzas <julian.bouzas@collabora.com>
Date: Thu, 15 Aug 2024 14:03:14 -0400
Subject: [PATCH 1/6] monitor: alsa: Add 'node.create-loopback' property

If enabled, the monitor will create a loopback filter for each ALSA node.
---
 src/config/wireplumber.conf   |  2 +-
 src/scripts/monitors/alsa.lua | 83 +++++++++++++++++++++++++++++++++++
 2 files changed, 84 insertions(+), 1 deletion(-)

diff --git a/src/config/wireplumber.conf b/src/config/wireplumber.conf
index 3fdf47b5..6da3701f 100644
--- a/src/config/wireplumber.conf
+++ b/src/config/wireplumber.conf
@@ -447,7 +447,7 @@ wireplumber.components = [
   {
     name = monitors/alsa.lua, type = script/lua
     provides = monitor.alsa
-    requires = [ support.client-context ]
+    requires = [ support.client-context, pw.node-factory.adapter ]
     wants = [ monitor.alsa.reserve-device ]
   }
   {
diff --git a/src/scripts/monitors/alsa.lua b/src/scripts/monitors/alsa.lua
index 8630e89d..1b1bef74 100644
--- a/src/scripts/monitors/alsa.lua
+++ b/src/scripts/monitors/alsa.lua
@@ -5,6 +5,7 @@
 --
 -- SPDX-License-Identifier: MIT
 
+LOOPBACK_ID_OFFSET = 128
 SPLIT_PCM_PARENT_OFFSET = 256
 SPLIT_PCM_OFFSET = 512
 
@@ -44,6 +45,79 @@ function shouldShowHdmiAlsaName (profile, properties, dev_props)
       properties["alsa.name"] ~= properties["device.profile.description"]
 end
 
+function CreateLoopback (properties)
+  local name = properties ["node.name"]
+  local desc = properties ["node.description"]
+  local device_media_class = properties ["media.class"]
+  local dev_id = properties ["device.id"]
+  local channels = properties ["audio.channels"]
+  local position = properties ["audio.position"]
+  local pri_driver = properties ["priority.driver"]
+  local pri_session = properties ["priority.session"]
+  local cpd = properties ["card.profile.device"]
+  local stream_media_class = nil
+  local stream_props = nil
+  local device_props = nil
+  local args = nil
+
+  -- Set stream media class
+  if device_media_class == "Audio/Sink" then
+    stream_media_class = "Stream/Output/Audio"
+  elseif device_media_class == "Audio/Source" then
+    stream_media_class = "Stream/Input/Audio"
+  else
+    return nil
+  end
+
+  -- Set stream props
+  stream_props = Json.Object {
+    ["node.name"] = string.format ("alsa_loopback_stream.%s", name),
+    ["node.description"] =
+        string.format ("ALSA internal stream for %s", desc),
+    ["media.class"] = stream_media_class,
+    ["audio.channels"] = channels,
+    ["audio.position"] = position,
+    ["alsa.loopback"] = true,
+    ["stream.dont-remix"] = true,
+    ["node.passive"] = true,
+    ["node.dont-fallback"] = true,
+    ["node.linger"] = true,
+    ["target.object"] = name
+  }
+
+  -- Set device props
+  device_props = Json.Object {
+    ["node.name"] = string.format ("alsa_loopback_device.%s", name),
+    ["node.description"] = desc,
+    ["media.class"] = device_media_class,
+    ["audio.channels"] = channels,
+    ["audio.position"] = position,
+    ["alsa.loopback"] = true,
+    ["device.id"] = dev_id,
+    ["card.profile.device"] = cpd,
+    ["priority.driver"] = tonumber(pri_driver) + 1,
+    ["priority.session"] = tonumber(pri_session) + 1
+  }
+
+  -- Set args
+  if device_media_class == "Audio/Sink" then
+    args = Json.Object {
+      ["playback.props"] = stream_props,
+      ["capture.props"] = device_props
+    }
+  elseif device_media_class == "Audio/Source" then
+    args = Json.Object {
+      ["playback.props"] = device_props,
+      ["capture.props"] = stream_props
+    }
+  else
+    return nil
+  end
+
+  -- Load loopback
+  return LocalModule("libpipewire-module-loopback", args:get_data(), {})
+end
+
 function createSplitPCMHWNode(dev_props, properties)
   local skip_keys = {
     "api.alsa.split.position", "card.profile.device", "device.profile.description",
@@ -449,9 +523,18 @@ function createNode(parent, id, obj_type, factory, properties)
         parent:store_managed_object(id, n)
       end
   end)
+
+  -- create the loopback node if requested
+  if cutils.parseBool (properties ["node.create-loopback"]) then
+    local loopback = CreateLoopback (properties)
+    parent:store_managed_object(LOOPBACK_ID_OFFSET + id, loopback)
+  end
 end
 
 function removeNode(parent, id)
+  -- remove loopback if any
+  parent:store_managed_object(LOOPBACK_ID_OFFSET + id, nil)
+
   local parent_id = tonumber(parent.properties["spa.object.id"])
   local ids = {id, SPLIT_PCM_PARENT_OFFSET + id, SPLIT_PCM_OFFSET + id}
 
-- 
2.55.0

